File: src/phoenix/sim2real/gate.py
"""The deploy-path safety gate ladder, as a pure function.

This module exists because the ladder it encodes previously lived inline in
``ros2_policy_node._control_step`` and therefore had **zero** test coverage:
every individual predicate in :mod:`phoenix.sim2real.safety` was well tested,
but the *composition* (which gate wins, what gets published when it fires,
and whether the abort latches) was not exercised by anything.

``evaluate_gates`` is a pure function of ``(snapshot, config, already_latched)``.
It touches no ROS, no ONNX, no clock and no ``self``, so it can be exhaustively
tested in CI, and it serves as the parity oracle for the native C++ port.

Ordering here is a literal transcription of the ladder as it ran in
``_control_step``. The one deliberate addition is the ``nan_in_imu`` gate: the
attitude gate below it compares ``abs(pitch) > threshold``, and ``abs(NaN)`` is
never greater than anything, so a non-finite quaternion previously slipped past
it and went on to poison the observation with NaN projected gravity.

The output contract is deliberately explicit. Two abort causes publish nothing
at all (``max_runtime`` and a previously-latched abort), and after any latch the
node must go silent permanently: re-broadcasting ``default_q`` every tick walked
the commanded pose against the robot's real posture, which caused motor fight
and a Jetson brownout. That silence is load-bearing behaviour, not an oversight.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

import numpy as np

from .safety import is_ready_to_command_motion, startup_state


class Outcome(Enum):
    """What the caller must do with the tick.

    The enum encodes the full output contract: whether to publish, what to
    publish, and whether to latch. It is not merely an error code.
    """

    #: Abort already latched on a previous tick. Publish nothing, ever again.
    SILENT = "silent"
    #: Latch now, publish nothing. Used only by ``max_runtime``, matching the
    #: pre-existing behaviour in which the latch fell through to the
    #: already-latched early return without emitting a pose.
    LATCH_SILENT = "latch_silent"
    #: Publish the default stand pose, do not latch. The only repeatedly
    #: publishing state; used while waiting for every topic to be seen once.
    PUBLISH_DEFAULT = "publish_default"
    #: Latch now and publish the default stand pose exactly once.
    LATCH_AND_PUBLISH_DEFAULT = "latch_and_publish_default"
    #: All gates passed. Run inference, apply the shield, clip and publish.
    RUN_POLICY = "run_policy"


@dataclass(frozen=True)
class GateConfig:
    """Thresholds governing the ladder. Every one has exactly one owner."""

    max_runtime_s: float
    estop_timeout_s: float
    sensor_timeout_s: float
    first_message_timeout_s: float
    pitch_rad: float
    roll_rad: float


@dataclass(frozen=True)
class SensorSnapshot:
    """Everything the ladder is allowed to look at, sampled once per tick.

    ``joint_pos`` / ``joint_vel`` are already remapped into policy joint order
    by the caller, because the remap is resolved by joint *name* and a
    positional copy would be a silent leg swap.
    """

    now_ns: int
    elapsed_s: float
    node_started_ns: int

    seen_estop: bool
    seen_imu: bool
    seen_joint_state: bool

    estop_last_ns: int | None
    estop_value: bool | None
    imu_last_ns: int | None
    joint_state_last_ns: int | None

    joint_pos: np.ndarray
    joint_vel: np.ndarray

    quat_xyzw: tuple[float, float, float, float]
    ang_vel: tuple[float, float, float]

    roll: float
    pitch: float


@dataclass(frozen=True)
class GateDecision:
    """The ladder's verdict for one tick."""

    outcome: Outcome
    #: Populated only for the latching outcomes. ``None`` otherwise.
    reason: str | None = None

    @property
    def latches(self) -> bool:
        return self.outcome in (
            Outcome.LATCH_SILENT,
            Outcome.LATCH_AND_PUBLISH_DEFAULT,
        )

    @property
    def publishes_default(self) -> bool:
        return self.outcome in (
            Outcome.PUBLISH_DEFAULT,
            Outcome.LATCH_AND_PUBLISH_DEFAULT,
        )


def evaluate_gates(
    snapshot: SensorSnapshot,
    config: GateConfig,
    *,
    already_latched: bool,
) -> GateDecision:
    """Evaluate the ladder. First gate to fire wins and short-circuits.

    Preserves the original ordering exactly. The ``already_latched`` flag is
    passed in rather than read from node state so that the function stays pure.
    """
    # Rank 1: runtime watchdog. Latches but emits nothing, because the
    # original code latched here and then hit the already-latched early
    # return on the very same tick.
    if snapshot.elapsed_s > config.max_runtime_s and not already_latched:
        return GateDecision(Outcome.LATCH_SILENT, "max_runtime")

    # Rank 0/1 continued: an abort latched on any previous tick, including
    # the out-of-band external estop, which latches between ticks.
    if already_latched:
        return GateDecision(Outcome.SILENT)

    # Rank 3: startup gate.
    startup, startup_reason = startup_state(
        seen_estop=snapshot.seen_estop,
        seen_imu=snapshot.seen_imu,
        seen_joint_state=snapshot.seen_joint_state,
        node_started_ns=snapshot.node_started_ns,
        now_ns=snapshot.now_ns,
        first_message_timeout_s=config.first_message_timeout_s,
    )
    if startup == "waiting":
        return GateDecision(Outcome.PUBLISH_DEFAULT)
    if startup == "abort":
        return GateDecision(
            Outcome.LATCH_AND_PUBLISH_DEFAULT,
            startup_reason or "first_message_timeout_unknown",
        )

    # Rank 2/4: estop-chain integrity and sensor freshness.
    ok, reason = is_ready_to_command_motion(
        now_ns=snapshot.now_ns,
        estop_last_ns=snapshot.estop_last_ns,
        estop_value=snapshot.estop_value,
        estop_timeout_s=config.estop_timeout_s,
        imu_last_ns=snapshot.imu_last_ns,
        joint_state_last_ns=snapshot.joint_state_last_ns,
        sensor_timeout_s=config.sensor_timeout_s,
    )
    if not ok:
        return GateDecision(Outcome.LATCH_AND_PUBLISH_DEFAULT, reason or "unknown_safety_gate")

    # Rank 4: joint-state validity.
    if not np.all(np.isfinite(snapshot.joint_pos)) or not np.all(np.isfinite(snapshot.joint_vel)):
        return GateDecision(Outcome.LATCH_AND_PUBLISH_DEFAULT, "nan_in_joint_state")

    # Rank 4: IMU validity. Must precede the attitude gate, which cannot
    # fire on NaN because abs(NaN) > threshold is False.
    if not np.all(np.isfinite(snapshot.quat_xyzw)) or not np.all(np.isfinite(snapshot.ang_vel)):
        return GateDecision(Outcome.LATCH_AND_PUBLISH_DEFAULT, "nan_in_imu")

    # Rank 6: attitude abort.
    if abs(snapshot.pitch) > config.pitch_rad or abs(snapshot.roll) > config.roll_rad:
        return GateDecision(
            Outcome.LATCH_AND_PUBLISH_DEFAULT,
            f"attitude pitch={snapshot.pitch:.2f} roll={snapshot.roll:.2f}",
        )

    # Rank 8: nothing seized the output; the policy may command motion.
    return GateDecision(Outcome.RUN_POLICY)

File: runtime/phoenix_core/src/gate.cpp
// Copyright 2026 Yusuf Guenena. MIT License.
// See gate.hpp. Line-for-line port of phoenix/sim2real/gate.py.
#include "phoenix_core/gate.hpp"

#include <cmath>

namespace phoenix_core
{

namespace
{

constexpr double kNsPerSec = 1e9;

double age_s(std::int64_t last_ns, std::int64_t now_ns) noexcept
{
  return static_cast<double>(now_ns - last_ns) / kNsPerSec;
}

bool all_finite(const JointArray & a) noexcept
{
  for (float v : a) {
    if (!std::isfinite(v)) {
      return false;
    }
  }
  return true;
}

bool all_finite(const Vec3d & a) noexcept
{
  for (double v : a) {
    if (!std::isfinite(v)) {
      return false;
    }
  }
  return true;
}

bool quat_finite(const QuatXYZW & q) noexcept
{
  return std::isfinite(q.x) && std::isfinite(q.y) && std::isfinite(q.z) && std::isfinite(q.w);
}

}  // namespace

bool estop_is_active(
  std::int64_t last_msg_ns, bool value, bool value_known, std::int64_t now_ns,
  double timeout_s) noexcept
{
  // Fail closed: never seen counts as asserted (safety.py:36-37).
  if (last_msg_ns == SensorSnapshot::kNeverSeen || !value_known) {
    return true;
  }
  if (age_s(last_msg_ns, now_ns) > timeout_s) {
    return true;
  }
  return value;
}

bool sensor_is_stale(std::int64_t last_msg_ns, std::int64_t now_ns, double timeout_s) noexcept
{
  if (last_msg_ns == SensorSnapshot::kNeverSeen) {
    return true;
  }
  return age_s(last_msg_ns, now_ns) > timeout_s;
}

GateDecision evaluate_gates(
  const SensorSnapshot & s, const GateConfig & c, bool already_latched) noexcept
{
  GateDecision d;

  // Rank 1: runtime watchdog. Latches but emits nothing, because the Python
  // latched here and then hit the already-latched early return on the same
  // tick. The asymmetry with every other abort is deliberate.
  if (s.elapsed_s > c.max_runtime_s && !already_latched) {
    d.outcome = Outcome::kLatchSilent;
    d.reason = AbortReason::kMaxRuntime;
    return d;
  }

  // Rank 0/1 continued: an abort latched on any previous tick, including the
  // out-of-band external estop, which latches between ticks.
  if (already_latched) {
    d.outcome = Outcome::kSilent;
    return d;
  }

  // Rank 3: startup gate (safety.py:167-211).
  const bool all_seen = s.seen_estop && s.seen_imu && s.seen_joint_state;
  if (!all_seen) {
    if (age_s(s.node_started_ns, s.now_ns) <= c.first_message_timeout_s) {
      d.outcome = Outcome::kPublishDefault;
      return d;
    }
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kFirstMessageTimeout;
    d.missing_estop = !s.seen_estop;
    d.missing_imu = !s.seen_imu;
    d.missing_joint_state = !s.seen_joint_state;
    return d;
  }

  // Rank 2/4: estop-chain integrity and sensor freshness
  // (is_ready_to_command_motion, safety.py:112-164). Reason ordering is part
  // of the contract: a missing publisher is distinguished from an asserted
  // estop, which is distinguished from a stale heartbeat, because the reason
  // string is what drives the lab post-mortem.
  if (s.estop_last_ns == SensorSnapshot::kNeverSeen || !s.estop_value_known) {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kEstopPublisherMissing;
    return d;
  }
  if (estop_is_active(
      s.estop_last_ns, s.estop_value, s.estop_value_known, s.now_ns, c.estop_timeout_s))
  {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = s.estop_value ? AbortReason::kExternalEstop : AbortReason::kEstopHeartbeatStale;
    return d;
  }
  if (s.imu_last_ns == SensorSnapshot::kNeverSeen ||
    s.joint_state_last_ns == SensorSnapshot::kNeverSeen)
  {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kSensorMissing;
    return d;
  }
  if (sensor_is_stale(s.imu_last_ns, s.now_ns, c.sensor_timeout_s) ||
    sensor_is_stale(s.joint_state_last_ns, s.now_ns, c.sensor_timeout_s))
  {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kSensorStale;
    return d;
  }

  // Rank 4: joint-state validity.
  if (!all_finite(s.joint_pos) || !all_finite(s.joint_vel)) {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kNanInJointState;
    return d;
  }

  // Rank 4: IMU validity. Must precede the attitude gate, which cannot fire on
  // NaN because abs(NaN) > threshold is false in both languages.
  if (!quat_finite(s.quat) || !all_finite(s.ang_vel)) {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kNanInImu;
    return d;
  }

  // Rank 6: attitude abort. Thresholds are asymmetric (0.8 pitch, 0.6 roll)
  // and the comparison is strictly greater than.
  if (std::fabs(s.pitch) > c.pitch_rad || std::fabs(s.roll) > c.roll_rad) {
    d.outcome = Outcome::kLatchAndPublishDefault;
    d.reason = AbortReason::kAttitude;
    return d;
  }

  // Rank 8: nothing seized the output; the policy may command motion.
  d.outcome = Outcome::kRunPolicy;
  return d;
}

}  // namespace phoenix_core
